Src/DependencyCollector/Shared/Implementation/SqlClientDiagnosticSourceListener.cs

```diff
@@ -0,0 +1,522 @@
+                        (Exception)ConnectionError.Exception.Fetch(evnt.Value),
+                        (long)ConnectionError.Timestamp.Fetch(evnt.Value));
+
+                    break;
+                }
+
+                case SqlBeforeCommitTransaction:
+                case SqlBeforeRollbackTransaction:
+                {
+                    this.operationStartTimestamps.TryAdd(
+                        (Guid)TransactionBefore.OperationId.Fetch(evnt.Value),
+                        (long)TransactionBefore.Timestamp.Fetch(evnt.Value));
+
+                    break;
+                }
+
+                case SqlAfterCommitTransaction:
+                case SqlAfterRollbackTransaction:
+                {
+                    this.OnAfterTransactionEvent(
+                        (Guid)TransactionAfter.OperationId.Fetch(evnt.Value),
+                        (string)TransactionAfter.Operation.Fetch(evnt.Value),
+                        (IsolationLevel)TransactionAfter.IsolationLevel.Fetch(evnt.Value),
+                        (SqlConnection)TransactionAfter.Connection.Fetch(evnt.Value),
+                        (long)TransactionAfter.Timestamp.Fetch(evnt.Value));
+
+                    break;
+                }
+
+                case SqlErrorCommitTransaction:
+                case SqlErrorRollbackTransaction:
+                {
+                    this.OnErrorTransactionEvent(
+                        (Guid)TransactionError.OperationId.Fetch(evnt.Value),
+                        (string)TransactionError.Operation.Fetch(evnt.Value),
+                        (IsolationLevel)TransactionError.IsolationLevel.Fetch(evnt.Value),
+                        (SqlConnection)TransactionError.Connection.Fetch(evnt.Value),
+                        (Exception)TransactionError.Exception.Fetch(evnt.Value),
+                        (long)TransactionError.Timestamp.Fetch(evnt.Value));
+
+                    break;
+                }
+            }
+        }
+
+        private static void InitializeTelemetry(ITelemetry telemetry, Guid operationId)
+        {
+            var activity = Activity.Current;
+
+            if (activity != null)
+            {
+                telemetry.Context.Operation.Id = activity.RootId;
+            }
+            else
+            {
+                telemetry.Context.Operation.Id = operationId.ToString("N");
+            }
+        }
+
+        private static void CorrelateErrorTelemetry(ExceptionTelemetry telemetry, Guid operationId)
+        {
+            telemetry.Context.Operation.ParentId = operationId.ToString("N");
+        }
+
+        private void OnAfterCommandEvent(
+            Guid operationId,
+            string operation,
+            Guid connectionId,
+            SqlCommand command,
+            IDictionary statistics,
+            long endTimestamp)
+        {
+            // Call DeriveDuration first to ensure we remove any start timestamp from operationStartTimestamps
```

Comment:
Could you please also add EventSource log for every callback with the message containing event name and operationId.
It helps a lot to debug things. 
@cijothomas ;)